src/main/java/org/opensearch/knn/quantization/factory/QuantizerRegistry.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.quantization.factory;
+
+import org.opensearch.knn.quantization.enums.QuantizationType;
+import org.opensearch.knn.quantization.enums.ScalarQuantizationType;
+import org.opensearch.knn.quantization.models.quantizationParams.QuantizationParams;
+import org.opensearch.knn.quantization.quantizer.Quantizer;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+/**
+ * The QuantizerRegistry class is responsible for managing the registration and retrieval
+ * of quantizer instances. Quantizers are registered with specific quantization parameters
+ * and type identifiers, allowing for efficient lookup and instantiation.
+ */
+final class QuantizerRegistry {
+
+    // Private constructor to prevent instantiation
+    private QuantizerRegistry() {}
+
+    // ConcurrentHashMap for thread-safe access
+    private static final Map<String, Supplier<? extends Quantizer<?, ?>>> registry = new ConcurrentHashMap<>();
+
+    /**
+     * Registers a quantizer with the registry.
+     *
+     * @param paramClass        the class of the quantization parameters
+     * @param quantizationType  the quantization type (e.g., VALUE_QUANTIZATION)
+     * @param sqType            the specific quantization subtype (e.g., ONE_BIT, TWO_BIT)
+     * @param quantizerSupplier a supplier that provides instances of the quantizer
+     * @param <P>               the type of quantization parameters
+     */
+    public static <P extends QuantizationParams> void register(
+        final Class<P> paramClass,
+        final QuantizationType quantizationType,
+        final ScalarQuantizationType sqType,
+        final Supplier<? extends Quantizer<?, ?>> quantizerSupplier
+    ) {
+        String identifier = createIdentifier(quantizationType, sqType);
+        // Ensure that the quantizer for this identifier is registered only once
+        registry.computeIfAbsent(identifier, key -> quantizerSupplier);
```

Comment:
What will happen in this case:
```
key1 -> QT1
key1 -> QT2
```
?

Reply:
Behavior of computeIfAbsent
* First Call with key1 -> QT1:
        If paramIdentifier (key1) is not present in the registry, the lambda function (key -> quantizerSupplier) will be called, and the result (QT1) will be inserted into the registry with the key (key1).
* Subsequent Call with key1 -> QT2:
        If paramIdentifier (key1) is already present in the registry, computeIfAbsent will not call the lambda function. Instead, it will return the existing value associated with the key (QT1), and QT2 will be ignored.